### PR TITLE
Move deny lint checks to script

### LINKED ISSUE #1551: Move `#![deny(...)]` from `lib.rs` to CI (script)
The denies are annoying during refactoring when one wants to iterate without bothering with unimportant stuff like unused import until the design is ready. Moving these to `-D` flags would improve the situation (now I have to patch `lib.rs` every time).

### PR DESCRIPTION
**Pull request by Harshil-Jani**
Closes #1551 
Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

### CODE REVIEW

**Comment by apoelstra** [comment]
[This article](https://matklad.github.io/2021/09/04/fast-rust-builds.html) which I am coincidentally reading right now suggests just putting `RUSTFLAGS=-D warnings` across the whole CI script rather than doing one run. (It also has a bunch of other easy to apply suggestions; search "CI Workflow"; but these are unrelated to this PR.)

OTOH I think `cargo clippy --all-targets --all-features` is likely to accomplish what we want, so I'm happy to ACK this as-is.

**Comment by apoelstra** [comment]
BTW it looks like CI is having some sort of network hiccup. I'll kick this in a few hours, or another day.

**Review by apoelstra — APPROVED** [review]
ACK 2ad5ba59bf44eee5100ea3ef928fdaf28a078945

**Comment by sanket1729** [comment]
Amongst all the lints that we are deleting, are we losing anything that the compiler would not warn us?

**Comment by apoelstra** [comment]
Good question. I think `missing_docs` is not a warning by default, so we should put a `warn(missing_docs)` back in. But the rest I'm pretty sure are warnings.

**Comment by Harshil-Jani** [comment]
Thanks @apoelstra for the Article (Added to reading list). I have updated the commit. Does it look good now ?

**Comment by apoelstra** [comment]
Could you copy the `if` syntax from `bitcoin/contrib/test.sh` which gates the existing `cargo clippy` invocations?

**Comment by apoelstra** [comment]
This is really interesting .... I'm surprised that we can run the examples in the root like this even though the examples actually live in `bitcoin/`. I wonder what would happen if multiple crates had examples with the same name.

@Harshil-Jani could you either:

* Add `-D warnings`to the existing `cargo clippy` lines and add only a single new `cargo clippy --all-targets --all-features` line; or
* Do what you've done, but remove the existing `cargo clippy` block because it's now redundant

File: contrib/test.sh [hunk 1]
```diff
@@ -2,6 +2,15 @@
...
 
+if [ "$DO_LINT" = true ]
+then
+    cargo clippy --all-features --all-targets -- -D warnings
+    cargo clippy --example bip32 -- -D warnings
+    cargo clippy --example handshake --features=rand-std -- -D warnings
+    cargo clippy --example ecdsa-psbt --features=bitcoinconsensus -- -D warnings
+    cargo clippy --example taproot-psbt --features=rand-std,bitcoinconsensus -- -D warnings
+fi
+
```
Comment (tcharding) [inline contrib/test.sh:13]: We already have these lines in the `bitcoin/contrib/test.sh` script, what is the purpose of adding them here?

Comment (apoelstra) [reply contrib/test.sh:13]: This ones have `-D warnings` added.

**Comment by Harshil-Jani** [comment]
> This is really interesting .... I'm surprised that we can run the examples in the root like this even though the examples actually live in `bitcoin/`. I wonder what would happen if multiple crates had examples with the same name.
> 
> @Harshil-Jani could you either:
> 
> * Add `-D warnings`to the existing `cargo clippy` lines and add only a single new `cargo clippy --all-targets --all-features` line; or
> * Do what you've done, but remove the existing `cargo clippy` block because it's now redundant

Going ahead with the second implementation.

File: contrib/test.sh (hunk 1 above)
Comment (tcharding) [reply contrib/test.sh:13]: We have `-D warnings` in `bitcoin/contrib/test.sh` also

**Comment by apoelstra** [comment]
This looks good to me. Interesting that the `--all-targets` line was already present in `bitcoin/contrib.sh`. It definitely looks like it belongs in the top-level contrib.sh. Thanks!

**Review by apoelstra — DISMISSED** [review]
ACK 98bfd529cce22a13beb62e3e1953793893a6d7f1

**Comment by tcharding** [comment]
I had a play with this. Looks like the explicit linting of the examples is unneeded, and git blame says I'm the goose that added those in the first place :)

As far as I can tell we need no changes to the CI scripts at all in this PR to achieve the closing of the issue we just remove the "coding convention" attributes as is done. The CI scripts in bitcoin, hashes, and internals already have calls to clippy as: `cargo clippy --all-features --all-targets -- -D warnings`.

**Comment by Harshil-Jani** [comment]
Thankyou @tcharding and @apoelstra for clarity and helping out. 

> I had a play with this. Looks like the explicit linting of the examples is unneeded, and git blame says I'm the goose that added those in the first place :)
> 
> As far as I can tell we need no changes to the CI scripts at all in this PR to achieve the closing of the issue we just remove the "coding convention" attributes as is done. The CI scripts in bitcoin, hashes, and internals already have calls to clippy as: `cargo clippy --all-features --all-targets -- -D warnings`.

**Review by apoelstra — APPROVED** [review]
ACK a1c3082319d6bb752b4e2144a0187d7f4a98a509

**Review by Kixunil — APPROVED** [review]
ACK a1c3082319d6bb752b4e2144a0187d7f4a98a509

**Pull request merged**